File: text_preprocessor.py
"""
Text Preprocessing Module
Handles text cleaning and preprocessing
"""

import re
from typing import List, Dict
import spacy
from config import SPACY_MODEL


class TextPreprocessor:
    """Handles text cleaning and preprocessing"""
    
    def __init__(self):
        self.nlp = spacy.load(SPACY_MODEL)
    
    def clean_text(self, text: str) -> str:
        """Clean and normalize text"""
        text = re.sub(r'\s+', ' ', text)
        text = re.sub(r'[^\w\s.,;:!?-]', '', text)
        text = text.strip()
        print("✓ Text cleaned")
        return text
    
    def sentence_tokenize(self, text: str) -> List[str]:
        """Split text into sentences"""
        doc = self.nlp(text)
        sentences = [sent.text.strip() for sent in doc.sents]
        print(f"✓ Text split into {len(sentences)} sentences")
        return sentences
    
    def preprocess(self, text: str) -> Dict:
        """Complete preprocessing pipeline"""
        cleaned = self.clean_text(text)
        sentences = self.sentence_tokenize(cleaned)
        
        return {
            'original': text,
            'cleaned': cleaned,
            'sentences': sentences
        }

File: backend/src/main/kotlin/io/github/juantoanxoup/kg/TextPreprocessor.kt
package io.github.juantoanxoup.kg

import edu.stanford.nlp.pipeline.CoreDocument
import org.slf4j.LoggerFactory

/** Output of [TextPreprocessor.preprocess]. */
data class Preprocessed(
    val original: String,
    val cleaned: String,
    val sentences: List<String>,
)

/** Text cleaning and sentence splitting (text_preprocessor.py). */
class TextPreprocessor {
    private val log = LoggerFactory.getLogger(TextPreprocessor::class.java)

    /** Collapses whitespace, removes characters outside word/space/`.,;:!?-`, trims. `(?U)` matches Python's Unicode `\w`. */
    fun cleanText(text: String): String {
        val cleaned =
            text
                .replace(WHITESPACE, " ")
                .replace(DISALLOWED, "")
                .trim()
        log.info("Text cleaned")
        return cleaned
    }

    fun sentenceTokenize(text: String): List<String> {
        val doc = CoreDocument(text)
        Nlp.sentenceSplitter.annotate(doc)
        val sentences = doc.sentences().map { it.text().trim() }
        log.info("Text split into {} sentences", sentences.size)
        return sentences
    }

    fun preprocess(text: String): Preprocessed {
        val cleaned = cleanText(text)
        return Preprocessed(original = text, cleaned = cleaned, sentences = sentenceTokenize(cleaned))
    }

    private companion object {
        val WHITESPACE = Regex("\\s+")
        val DISALLOWED = Regex("(?U)[^\\w\\s.,;:!?-]")
    }
}
